core/imt/src/RTaskArena.cxx

```diff
@@ -0,0 +1,146 @@
+   if(stat(quotaFile.c_str(), &buffer) == 0) {
+      f.open(quotaFile);
+      float cfs_quota;
+      f>>cfs_quota;
+      f.close();
+      if(cfs_quota > 0) {
+         std::string periodFile("/sys/fs/cgroup/cpuacct/cpu.cfs_period_us");
+         f.open(periodFile);
+         float cfs_period;
+         f>>cfs_period;
+         f.close();
+         return static_cast<int>(std::ceil(cfs_quota/cfs_period));
+      }
+   }
+#endif
+   return std::thread::hardware_concurrency();
+}
+
+namespace ROOT{
+namespace Internal {
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Initializes the tbb::task_arena within RTaskArenaWrapper
+///
+/// * Can't be reinitialized
+/// * Checks for CPU bandwidth control and avoids oversubscribing
+/// * If no BC in place and maxConcurrency<1, defaults to the default tbb number of threads,
+/// which is CPU affinity aware
+////////////////////////////////////////////////////////////////////////////////
+RTaskArenaWrapper::RTaskArenaWrapper(unsigned maxConcurrency): fTBBArena(new tbb::task_arena{})
+{
+   if (!fTBBArena->is_active()) {
+      unsigned tbbDefaultNumberThreads = fTBBArena->max_concurrency(); // not initialized, automatic state
+      maxConcurrency = maxConcurrency > 0 ? std::min(maxConcurrency, tbbDefaultNumberThreads) : tbbDefaultNumberThreads;
+      unsigned bcCpus = LogicalCPUBandwithControl();
+      if (maxConcurrency>bcCpus) {
+         Warning("RTaskArenaWrapper", "CPU Bandwith Control Active. Proceeding with %d threads accordingly",
+            bcCpus);
+         maxConcurrency = bcCpus;
+      }
+      fTBBArena->initialize(maxConcurrency);
+      fNWorkers = maxConcurrency;
+      ROOT::EnableThreadSafety();
+   } else {
+      unsigned current = fTBBArena->max_concurrency();
+      if (maxConcurrency && (current != maxConcurrency)) {
+         Warning("RTaskArenaWrapper", "There's already an active task arena. Proceeding with the current %d threads",
+            current);
+      }
+   }
+}
+
+RTaskArenaWrapper::~RTaskArenaWrapper()
+{
+   fNWorkers = 0u;
+}
+
+unsigned RTaskArenaWrapper::fNWorkers = 0u;
+
+unsigned RTaskArenaWrapper::TaskArenaSize()
+{
+   return fNWorkers;
+}
+////////////////////////////////////////////////////////////////////////////////
+/// Provides access to the wrapped tbb::task_arena
+////////////////////////////////////////////////////////////////////////////////
+std::unique_ptr<tbb::task_arena> &RTaskArenaWrapper::Access()
+{
+   return fTBBArena;
+}
+
+
+std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena(unsigned maxConcurrency)
+{
+   static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
+   if (weak_GTAWrapper.expired()) {
+      std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> shared_GTAWrapper(new ROOT::Internal::RTaskArenaWrapper(maxConcurrency));
+      weak_GTAWrapper = shared_GTAWrapper;
+      return weak_GTAWrapper.lock();
+   }
+
+   return weak_GTAWrapper.lock();
```

Comment:
Potential race condition. The the lifetime of the shared pointer pointed to by `weak_GTAWrapper` could expire between the call to `expired()` and the last `lock()`. Try to lock directly and inspect the result:

```
std::shared_ptr<ROOT::Internal::RTaskArenaWrapper> GetGlobalTaskArena(unsigned maxConcurrency)
{
  static std::weak_ptr<ROOT::Internal::RTaskArenaWrapper> weak_GTAWrapper;
  if (auto sp = weak_GTAWrapper.lock())
    return sp;
  auto sp = std::make_shared<ROOT::Internal::RTaskArenaWrapper>(maxConcurrency);
  weak_GTAWrapper = sp;
  return sp;
}
```
And prefer `std::make_shared` instead of repeating yourself.

Reply:
Yeah, this one we knew and decided to leave as is for the minimal chance of that happening You will have to call it concurrently from at least two different threads explicitly managed by the user, assuming you didn't initialize any ROOT MT class before.

Reply:
Are you serious? This is a potential bug and the fix is easy. There is really no reason to not apply it. It takes you 5min to do so and prevent a potential 3h or more debugging session of someone in the future.

Reply:
Sorry, I didn't look properly at your code. I thought we were talking about adding std::locks and making it a critical section. Nice solution, didn't think of that.